addons/website_sale/controllers/cart.py

```diff
@@ -0,0 +1,343 @@
+            elif revive == 'merge':
+                abandoned_order.order_line.write({'order_id': request.session['sale_order_id']})
+                abandoned_order.action_cancel()
+            elif abandoned_order.id != request.session.get('sale_order_id'):  # abandoned cart found, user have to choose what to do
+                values.update({'id': abandoned_order.id, 'access_token': abandoned_order.access_token})
+
+        values.update({
+            'website_sale_order': order,
+            'date': fields.Date.today(),
+            'suggested_products': [],
+        })
+        if order:
+            order.order_line.filtered(lambda sol: sol.product_id and not sol.product_id.active).unlink()
+            values['suggested_products'] = order._cart_accessories()
+            values.update(self._get_express_shop_payment_values(order))
+
+        values.update(self._cart_values(**post))
+        return request.render('website_sale.cart', values)
+
+    def _get_express_shop_payment_values(self, order, **kwargs):
+        payment_form_values = CustomerPortal._get_payment_values(
+            self, order, website_id=request.website.id, is_express_checkout=True
+        )
+        payment_form_values.update({
+            'payment_access_token': payment_form_values.pop('access_token'),  # Rename the key.
+            'minor_amount': payment_utils.to_minor_currency_units(
+                order.amount_total, order.currency_id
+            ),
+            'merchant_name': request.website.name,
+            'transaction_route': f'/shop/payment/transaction/{order.id}',
+            'express_checkout_route': WebsiteSale._express_checkout_route,
+            'landing_route': '/shop/payment/validate',
+            'payment_method_unknown_id': request.env.ref('payment.payment_method_unknown').id,
+            'shipping_info_required': order._has_deliverable_products(),
+            'delivery_amount': payment_utils.to_minor_currency_units(
+                order.order_line.filtered(lambda l: l.is_delivery).price_total, order.currency_id
+            ),
+            'shipping_address_update_route': WebsiteSale._express_checkout_delivery_route,
+        })
+        if request.website.is_public_user():
+            payment_form_values['partner_id'] = -1
+        return payment_form_values
+
+    def _cart_values(self, **post):
+        """
+        This method is a hook to pass additional values when rendering the 'website_sale.cart' template (e.g. add
+        a flag to trigger a style variation)
+        """
+        return {}
+
+    @route(
+        route='/shop/cart/update',
+        type='jsonrpc',
+        auth='public',
+        methods=['POST'],
+        website=True,
+        sitemap=False
+    )
+    def update_cart(
+        self,
+        line_id,
+        product_id,
+        quantity,
+    ):
+        if not line_id:
+            return  # Ensures this method is only used from the cart page.
```

Comment:
Loyalty logic aggregates discount on different taxes in a new line, which doesn't have an id.
If you try to remove the discount line, this return nothing, and the js raises :eyes: